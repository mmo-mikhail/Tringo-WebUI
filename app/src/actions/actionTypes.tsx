export enum destinationActionType {
    FETCH_DESTINATION_START = "FETCH_DESTINATION_START",
    FETCH_DESTINATION_SUCCESS = "FETCH_DESTINATION_SUCCESS",
    FETCH_DESTINATION_FAIL = "FETCH_DESTINATION_FAIL",
}