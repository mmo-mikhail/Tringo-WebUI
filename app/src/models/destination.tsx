import { Record } from "immutable";
import IBaseModel from "./BaseModelInterface";
import _ from "lodash";

export interface IDestination {
	lat?: number;
	lng?: number;
	price?: number;
	cityName?: string;
}

export interface IDestinationStore extends IDestination, IBaseModel { }

const defaultValues: IDestinationStore = {
	lat: undefined,
	lng: undefined,
	price: undefined,
	cityName: undefined,
	isLoading: false,
	error: null,
};

export default class Destination extends Record(defaultValues) {
	constructor(js?: any ) {
		const additionalFields: IBaseModel = {
			isLoading: false,
			error: null,
		};

		js ? super(_.merge(js, additionalFields)) : super();
	}
}