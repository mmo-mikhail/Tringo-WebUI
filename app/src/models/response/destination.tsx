export interface IDestination {
    lat: number;
    lng: number;
    price: number;
    cityName: string;
    destAirportCode: string;
    personalPriorityIdx: number;
    flightDates: IFlightDates;
}

export interface IFlightDates {
    flightMonthidx: number;
}
