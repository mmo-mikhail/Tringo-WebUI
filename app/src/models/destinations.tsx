import { Record } from "immutable";
import IBaseModel from "./BaseModelInterface";
import { Destination } from "./destination";
import _ from "lodash";

export interface IDestinations {
	destinations: Destination[] | null;
}

export interface IDestinationsStore extends IDestinations, IBaseModel { }

const defaultValues: IDestinationsStore = {
	destinations: null,
	isLoading: false,
	error: null,
};

export class DestinationsState extends Record(defaultValues) {
	constructor(js?: any) {
		const additionalFields: IBaseModel = {
			isLoading: false,
			error: null,
		};

		js ? super(_.merge(js, additionalFields)) : super();
	}
}