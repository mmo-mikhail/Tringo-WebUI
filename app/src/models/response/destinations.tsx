import { Record } from 'immutable';
import IBaseModel from '../BaseModelInterface';
import { IDestination } from './destination';
import _ from 'lodash';

export interface IDestinations {
    destinations: IDestination[] | null;
}

export interface IDestinationsStore extends IDestinations, IBaseModel {}

const defaultValues: IDestinationsStore = {
    destinations: null,
    isLoading: false,
    error: null
};

export class DestinationsState extends Record(defaultValues) implements IDestinationsStore {
    constructor(js?: any) {
        const additionalFields: IBaseModel = {
            isLoading: false,
            error: null
        };

        js ? super(_.merge(js, additionalFields)) : super();
    }
}
