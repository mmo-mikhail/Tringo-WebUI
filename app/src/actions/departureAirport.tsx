import { departureAirportActionType } from './actionTypes';
import { IDepartureAirport } from 'models/response/departureAirport';
import { DepartureAirportRequest, IDepartureAirportRequestAction } from '../models/request/departureAirportRequest';

export const fetchDepartureAirportStart = (model: DepartureAirportRequest): IDepartureAirportRequestAction => {
    return {
        type: departureAirportActionType.FETCH_DEPARTUREAIRPORT_START,
        model
    };
};

export const fetchDepartureAirportClear = (model: DepartureAirportRequest): IDepartureAirportRequestAction => {
    return {
        type: departureAirportActionType.FETCH_DEPARTUREAIRPORT_CLEAR,
        model
    };
};

export const fetchDepartureAirportSuccess = (departureAirport: IDepartureAirport) => {
    return {
        type: departureAirportActionType.FETCH_DEPARTUREAIRPORT_SUCCESS,
        departureAirport
    };
};

export const fetchDepartureAirportFail = (error: string) => {
    return {
        type: departureAirportActionType.FETCH_DEPARTUREAIRPORT_FAIL,
        error
    };
};
