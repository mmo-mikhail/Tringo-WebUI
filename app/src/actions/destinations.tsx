import { destinationActionType } from './actionTypes';
import { IDestination } from '../models/response/destination';
import {
    FlightDestinationRequest,
    IFlightsRequestAction
} from '../models/request/flightDestinationRequest';

export const fetchDestinationsStart = (
    model: FlightDestinationRequest
): IFlightsRequestAction => {
    return {
        type: destinationActionType.FETCH_DESTINATION_START,
        model
    };
};

export const fetchDestinationsSuccess = (destinations: IDestination[]) => {
    return {
        type: destinationActionType.FETCH_DESTINATION_SUCCESS,
        destinations
    };
};

export const fetchDestinationsFail = (error: string) => {
    return {
        type: destinationActionType.FETCH_DESTINATION_FAIL,
        error
    };
};
