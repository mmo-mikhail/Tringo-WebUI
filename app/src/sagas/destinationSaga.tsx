import { put, call } from 'redux-saga/effects';

import axios from './../utils/axious';
import * as destinationAction from '../actions/destinations';
import { IFlightsRequestAction } from '../models/request/flightDestinationRequest';

export function* fetchDestinationsSaga(action: IFlightsRequestAction) {
    //console.log("fetchDestinationsSaga");
    //console.log(action.model);
    try {
        const response = yield call(
            axios.get,
            '/api/v1/flights/GetDestinationPrice'
            //'/api/v1/flights/GetDestinationPrices', // note the 's' at the end
            //{
            //    data: action.model,
            //    headers: {
            //        'Content-Type': 'application/json'
            //    }
            //}
        );

        yield put(destinationAction.fetchDestinationsSuccess(response.data));
    } catch (error) {
        yield put(destinationAction.fetchDestinationsFail(error.message));
    }
}
