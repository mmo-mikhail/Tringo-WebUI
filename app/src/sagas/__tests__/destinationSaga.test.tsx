import { put } from "redux-saga/effects";
import { fetchDestinationsSaga } from "../destinationSaga";
import * as actions from "../../actions/destinations";
import { IDestination } from "../../models/destination";

describe("fetchDestinationsSaga", () => {
	let fetchDestinationGenerator: any;

	beforeEach(() => {
		fetchDestinationGenerator = fetchDestinationsSaga();

		const callDescriptor = fetchDestinationGenerator.next().value;
		expect(callDescriptor).toMatchSnapshot();
	});

	it("should dispatch the fetchDestinationsSuccess action if it requests the data successfully", () => {
		interface IResponse {
			data: IDestination[];
		}

		const response: IResponse = {
			data: [],
		};

		const putDescriptor = fetchDestinationGenerator.next(response).value;
		expect(putDescriptor).toEqual(put(actions.fetchDestinationsSuccess(response.data)));
	});

	it("should dispatch the fetchDestinationsFail action if the response errors", () => {
		const error = new Error("Some error");
		const putDescriptor = fetchDestinationGenerator.throw(error).value;
		expect(putDescriptor).toEqual(put(actions.fetchDestinationsFail(error.message)));
	});
});