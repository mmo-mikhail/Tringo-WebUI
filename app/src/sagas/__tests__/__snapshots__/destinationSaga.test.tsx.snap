// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchDestinationsSaga should dispatch the fetchDestinationsFail action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/api/v1/flights/GetDestinationPrices",
      FlightDestinationRequest {
        "budget": Budget {
          "max": 1000,
          "min": 0,
        },
        "dates": DatesInput {
          "dateFrom": null,
          "dateUntil": null,
          "uncertainDates": UncertainDates {
            "duration": 1,
            "monthIdx": 10,
          },
        },
        "departureAirportId": "MEL",
        "searchArea": MapArea {
          "lat": -12.21,
          "lng": 66.6,
          "radius": 199,
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`fetchDestinationsSaga should dispatch the fetchDestinationsSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/api/v1/flights/GetDestinationPrices",
      FlightDestinationRequest {
        "budget": Budget {
          "max": 1000,
          "min": 0,
        },
        "dates": DatesInput {
          "dateFrom": null,
          "dateUntil": null,
          "uncertainDates": UncertainDates {
            "duration": 1,
            "monthIdx": 10,
          },
        },
        "departureAirportId": "MEL",
        "searchArea": MapArea {
          "lat": -12.21,
          "lng": 66.6,
          "radius": 199,
        },
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
